File: lib/web/removeLocalePrefix.ts
export function removeLocalePrefix(urlOrPathname: string, locales: string | string[]): string {
	const localeLists = Array.isArray(locales) ? locales : [locales];

	if (localeLists.includes(urlOrPathname.replace(/^\//, ''))) {
		return '';
	}

	if (urlOrPathname === '/') {
		return '/';
	}

	const joinLocaleLists = localeLists.join('|');

	try {
		const urlObj = new URL(urlOrPathname);

		if (urlObj.pathname === '/' || localeLists.includes(urlObj.pathname.replace(/^\//, ''))) {
			return urlObj.origin;
		}

		return `${urlObj.origin}${urlObj.pathname.replace(new RegExp(`^/(${joinLocaleLists})/`), '/')}`.replace(
			/\/$/,
			''
		);
	} catch {
		let realPathname = urlOrPathname;

		if (!realPathname.startsWith('/')) {
			realPathname = `/${realPathname}`;
		}

		if (realPathname.endsWith('/')) {
			realPathname = realPathname.replace(/\/$/, '');
		}

		if (new RegExp(`^/(${joinLocaleLists})$`).test(realPathname)) {
			return '/';
		}

		return realPathname.replace(new RegExp(`^/(${joinLocaleLists})/`), '/');
	}
}

File: packages/python/qsu/web/removeLocalePrefix.py
import re
from typing import List, Union
from urllib.parse import urlsplit


def removeLocalePrefix(urlOrPathname: str, locales: Union[str, List[str]]) -> str:
	localeLists = locales if isinstance(locales, list) else [locales]

	if re.sub(r'^/', '', urlOrPathname) in localeLists:
		return ''

	if urlOrPathname == '/':
		return '/'

	joinLocaleLists = '|'.join(localeLists)

	parsed = urlsplit(urlOrPathname)

	if parsed.scheme and parsed.netloc:
		origin = f'{parsed.scheme}://{parsed.netloc}'
		pathname = parsed.path if parsed.path else '/'

		if pathname == '/' or re.sub(r'^/', '', pathname) in localeLists:
			return origin

		newPathname = re.sub(rf'^/({joinLocaleLists})/', '/', pathname)
		return re.sub(r'/$', '', f'{origin}{newPathname}')

	realPathname = urlOrPathname

	if not realPathname.startswith('/'):
		realPathname = f'/{realPathname}'

	if realPathname.endswith('/'):
		realPathname = re.sub(r'/$', '', realPathname)

	if re.match(rf'^/({joinLocaleLists})$', realPathname):
		return '/'

	return re.sub(rf'^/({joinLocaleLists})/', '/', realPathname)
